src/core/midi/MidiJack.cpp

```diff
@@ -0,0 +1,243 @@
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "MidiJack.h"
+
+#ifdef LMMS_HAVE_JACK
+
+#include <QCompleter>
+#include <QDirModel>
+#include <QLabel>
+#include <QLineEdit>
+#include <QMessageBox>
+#include <QTranslator>
+
+#ifdef LMMS_HAVE_STDLIB_H
+#include <stdlib.h>
+#endif
+
+#include "ConfigManager.h"
+#include "gui_templates.h"
+#include "GuiApplication.h"
+#include "Engine.h"
+#include "Mixer.h"
+#include "MainWindow.h"
+
+/* callback functions for jack */
+static int JackMidiProcessCallback(jack_nframes_t nframes, void *arg)
+{
+	MidiJack *jmd = (MidiJack *)arg;
+
+	if (nframes <= 0)
+		return (0);
+
+	jmd->JackMidiRead(nframes);
+	jmd->JackMidiWrite(nframes);
+
+	return (0);
+}
+
+static void JackMidiShutdown(void *arg)
+{
+	// TODO: support translations here
+	const QString mess_short = "JACK server down";
+	const QString mess_long = "The JACK server seems to have been shutdown.";
+	QMessageBox::information( gui->mainWindow(), mess_short, mess_long );
+}
+
+MidiJack::MidiJack() :
+	MidiClientRaw(),
+	m_input_port( NULL ),
+	m_output_port( NULL ),
+	m_quit( false )
+{
+	// if jack is used for audio then we share the connection
+	// AudioJack creates and maintains the jack connection
+	// and also handles the callback, we pass it our address
+	// so that we can also process during the callback
+
+	if(Engine::mixer()->audioDevName() == AudioJack::name() )
+	{
+		// if a jack connection has been created for audio we use that
+		m_jackAudio = dynamic_cast<AudioJack*>(Engine::mixer()->audioDev())->addMidiClient(this);
+	}else{
+		m_jackAudio = NULL;
+		m_jackClient = jack_client_open(probeDevice().toLatin1().data(),
+										JackNoStartServer, NULL);
+
+		if(m_jackClient)
+		{
+			jack_set_process_callback(m_jackClient,
+							JackMidiProcessCallback, this);
+			jack_on_shutdown(m_jackClient,
+							JackMidiShutdown, 0);
+		}
+	}
+
+	if(jackClient())
+	{
+		m_output_port = jack_port_register(
+				jackClient(), "midi_TX", JACK_DEFAULT_MIDI_TYPE,
+				JackPortIsOutput, 0);
+
+		m_input_port = jack_port_register(
+				jackClient(), "midi_RX", JACK_DEFAULT_MIDI_TYPE,
+				JackPortIsInput, 0);
+
+		if(jack_activate(jackClient()) == 0 )
+		{
+			// only start thread, if we have an active jack client.
+			start( QThread::LowPriority );
+		}
+	}
+}
+
+MidiJack::~MidiJack()
+{
+	if(jackClient())
+	{
+		if( jack_port_unregister( jackClient(), m_input_port) != 0){
+			printf("Failed to unregister jack midi input\n");
+		}
+
+		if( jack_port_unregister( jackClient(), m_output_port) != 0){
+			printf("Failed to unregister jack midi output\n");
+		}
+
+		if(m_jackClient)
+		{
+			// an m_jackClient means we are handling the jack connection
+			if( jack_deactivate(m_jackClient) != 0){
+				printf("Failed to deactivate jack midi client\n");
+			}
+
+			if( jack_client_close(m_jackClient) != 0){
+				printf("Failed close jack midi client\n");
+			}
+		}
+	}
+	if( isRunning() )
+	{
+		m_quit = true;
+		wait( 1000 );
```

Comment:
That is in the destructor, set m_quit to true, wait a little for thread to stop then terminate, that  code was copied from MidiOss.cpp and is also used in MidiAlsaRaw.cpp